lib/subghz/subghz_keystore.c

```diff
@@ -315,3 +325,268 @@ SubGhzKeyArray_t* subghz_keystore_get_data(SubGhzKeystore* instance) {
+
+        } while(ret > 0 && result);
+
+        flipper_file_close(output_flipper_file);
+        flipper_file_free(output_flipper_file);
+
+        furi_hal_crypto_store_unload_key(SUBGHZ_KEYSTORE_FILE_ENCRYPTION_KEY_SLOT);
+
+        if(!result) break;
+
+        encrypted = true;
+    } while(0);
+
+    flipper_file_close(input_flipper_file);
+    flipper_file_free(input_flipper_file);
+
+    free(encrypted_line);
+
+    furi_record_close("storage");
+
+    return encrypted;
+}
+
+bool subghz_keystore_raw_get_data(const char* file_name, size_t bias, uint8_t* data, size_t len) {
+
+    bool result = false;
+    uint8_t iv[16];
+    uint32_t version;
+    SubGhzKeystoreEncryption encryption;
+
+    string_t str_temp;
+    string_init(str_temp);
+
+    Storage* storage = furi_record_open("storage");
+    char* decrypted_line = furi_alloc(SUBGHZ_KEYSTORE_FILE_DECRYPTED_LINE_SIZE);
+
+    FlipperFile* flipper_file = flipper_file_alloc(storage);
+    do {
+        if(!flipper_file_open_read(flipper_file, file_name)) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Unable to open file for read: %s", file_name);
+            break;
+        }
+        if(!flipper_file_read_header(flipper_file, str_temp, &version)) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Missing or incorrect header");
+            break;
+        }
+        if(!flipper_file_read_uint32(flipper_file, "Encryption", (uint32_t*)&encryption)) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Missing encryption type");
+            break;
+        }
+
+        if(strcmp(string_get_cstr(str_temp), SUBGHZ_KEYSTORE_FILE_RAW_TYPE) != 0 ||
+           version != SUBGHZ_KEYSTORE_FILE_VERSION) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Type or version mismatch");
+            break;
+        }
+
+        File* file = flipper_file_get_file(flipper_file);
+        if(encryption != SubGhzKeystoreEncryptionAES256) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Unknown encryption");
+            break;
+        }
+
+        if(bias < 16) {
+            if(!flipper_file_read_hex_array(flipper_file, "IV", iv, 16)) {
+                FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Missing IV");
+                break;
+            }
+            subghz_keystore_mess_with_iv(iv);
+        }
+
+        if(!flipper_file_read_string(flipper_file, "Encrypt_data", str_temp)) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Missing Encrypt_data");
+            break;
+        }
+
+        size_t bufer_size = (((len+bias) / 16) + 1) * 32;
+        furi_assert(SUBGHZ_KEYSTORE_FILE_DECRYPTED_LINE_SIZE >= bufer_size * 2);
+
+        char buffer[bufer_size];
+        size_t ret = 0;
+        bool decrypted = true;
+        //skip the end of the previous line "\n"
+        storage_file_read(file, buffer, 1);
+
+        size_t size = storage_file_size(file);
+        size -= storage_file_tell(file);
+        if(size < (bias * 2 + len * 2)) {
+            FURI_LOG_E(SUBGHZ_KEYSTORE_TAG, "Went outside the file");
```

Comment:
`"Went outside the file"` -> `"Seek position exceeds file size"`

Reply:
ok